packages/flutter/lib/src/widgets/router.dart

```diff
@@ -0,0 +1,1327 @@
+      });
+    } on JsonUnsupportedObjectError catch (error) {
+      FlutterError.reportError(
+        FlutterErrorDetails(
+          exception:
+          'The RouteInformation.state must be json serializable.\n' +
+            error.toString()),
+      );
+    }
+    return result;
+  }
+
+  static RouteInformation _deserializeRouteInformation(String serialized) {
+    final dynamic json = jsonDecode(serialized);
+    if (json is String)
+      return RouteInformation(location: json);
+    assert(json is Map<String, dynamic>);
+    final Map<String, dynamic> map = json as Map<String, dynamic>;
+    assert(map.containsKey('location') && map.containsKey('state'));
+    return RouteInformation(
+      location: json['location'] as String,
+      state: json['state'],
+    );
+  }
+
+  @override
+  void routerReportsNewRouteInformation(RouteInformation routeInformation) {
+    final String serializedRouteInformation = _serializeRouteInformation(routeInformation);
+    RouteNotificationMessages.maybeNotifyRouteChange(
+      serializedRouteInformation,
+      _currentSerializedRouteInformationInEngine
+    );
+    _currentSerializedRouteInformationInEngine = serializedRouteInformation;
+    _value = routeInformation;
+  }
+
+  void _updateAndNotify(RouteInformation newValue) {
+    assert(_value != newValue);
+    _value = newValue;
+    notifyListeners();
+  }
+
+  @override
+  RouteInformation get value => _value;
+  RouteInformation _value;
+  set value(RouteInformation newValue) {
+    if (_value == newValue)
+      return;
+    _updateAndNotify(newValue);
+    if (_currentSerializedRouteInformationInEngine != _serializeRouteInformation(newValue))
+      routerReportsNewRouteInformation(newValue);
+  }
+
+  void _platformReportsNewRouteInformation(RouteInformation routeInformation) {
+    if (_value == routeInformation)
+      return;
+    _updateAndNotify(routeInformation);
+    // We can't use the routerReportsNewRouteInformation because we don't want
+    // to report it back to the engine again.
+    _currentSerializedRouteInformationInEngine = _serializeRouteInformation(routeInformation);
+  }
+
+  @override
+  void addListener(VoidCallback listener) {
+    if (!hasListeners)
+      WidgetsBinding.instance.addObserver(this);
+    super.addListener(listener);
+  }
+
+  @override
+  void removeListener(VoidCallback listener) {
+    super.removeListener(listener);
+    if (!hasListeners)
+      WidgetsBinding.instance.removeObserver(this);
+  }
+
+  @override
+  void dispose() {
+    // In practice, this will rarely be called. We assume that the listeners
+    // will be added and removed in a coherent fashion such that when the object
+    // is no longer being used, there's no listener, and so it will get garbage
+    // collected.
+    if (hasListeners)
+      WidgetsBinding.instance.removeObserver(this);
+    super.dispose();
+  }
+
+  @override
+  Future<bool> didPushRoute(String route) async {
```

Comment:
Doesn't didPushRoute just give you a route name instead of a complex serialized object? Do we need new API to obtain a RouteInformation Object from the engine? Or maybe this method should have two params: location and state.